### PR TITLE
fix: add a check for / suffix for basyx external url in registry integration

### LINKED ISSUE #414: Add a check for "/" suffix for basyx external url in registry integration
Currently, the base URL for basyx.externalurl is not working when the context path is set and basyx.externalurl doesn't contain a leading "/" in the Registry integration scenario.

## Rules:

- Add a check if the basyx.externalurl has an additional context path set.
- If it is set then add a "/" at the end.

## Entry Points 

[1] [basyx-java-server-sdk/basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/AasDescriptorFactory.java at main · eclipse-basyx/basyx-java-server-sdk (github.com)](https://github.com/eclipse-basyx/basyx-java-server-sdk/blob/main/basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/AasDescriptorFactory.java)
[2] [basyx-java-server-sdk/basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/SubmodelDescriptorFactory.java at main · eclipse-basyx/basyx-java-server-sdk (github.com)](https://github.com/eclipse-basyx/basyx-java-server-sdk/blob/main/basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/main/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/SubmodelDescriptorFactory.java)

## Acceptance Criteria 

- [ ] The AAS/SM endpoint inside the descriptor is correct when the registry integration feature is set and basyx.externalurl has a context path set without the "/" suffix

### PR DESCRIPTION
**Pull request by ShehriyarShariq-Fraunhofer**
Added a trailing slash in case one was missing for the basyx external url in registry integration.

Closes #414

### CODE REVIEW

**Comment by ShehriyarShariq-Fraunhofer** [comment]
@mateusmolina-iese

File: basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        String baseURLWithSlash = AAS_REPO_URL + "/";
+
+        assertEquals(baseURLWithSlash + AAS_REPOSITORY_PATH_WITHOUT_SLASH, AasDescriptorFactory.createAasRepositoryUrl(baseURLWithSlash));
+    }
+	
+	@Test
+    public void testUrlWithTrailingSlashAndPathWithLeadingSlash() {
+        String baseURLWithSlash = AAS_REPO_URL + "/";
+
+        assertEquals(baseURLWithSlash + AAS_REPOSITORY_PATH_WITHOUT_SLASH, AasDescriptorFactory.createAasRepositoryUrl(baseURLWithSlash));
```
Comment (mateusmolina-iese) [inline basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java:25]: The assertion is wrongly defined

Comment (mateusmolina-iese) [inline basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java:25]: The assertion is wrongly defined

File: basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Test
+    public void testUrlWithoutTrailingSlashAndPathWithoutLeadingSlash() {
+        String baseURLWithoutSlash = AAS_REPO_URL;
+
+        assertEquals(baseURLWithoutSlash + AAS_REPOSITORY_PATH_WITH_SLASH , AasDescriptorFactory.createAasRepositoryUrl(baseURLWithoutSlash));
```
Comment (mateusmolina-iese) [inline basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java:32]: The assertion is wrongly defined

File: basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class TestAasDescriptorFactory {
+
+	private static final String AAS_REPO_URL = "http://localhost:8081";
+
+	private static final String AAS_REPOSITORY_PATH_WITH_SLASH = "/shells";
+	private static final String AAS_REPOSITORY_PATH_WITHOUT_SLASH = "shells";
```
Comment (mateusmolina-iese) [inline basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/TestAasDescriptorFactory.java:12]: Either stick to building the withSlash/withoutSlash for both path and repo_url in the test or define them consistenly as final vars; but don't mix both approaches.

File: basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        String baseURLWithSlash = SUBMODEL_REPO_URL + "/";
+
+        assertEquals(baseURLWithSlash + SUBMODEL_REPOSITORY_PATH_WITHOUT_SLASH, SubmodelDescriptorFactory.createSubmodelRepositoryUrl(baseURLWithSlash));
+    }
+	
+	@Test
+    public void testUrlWithTrailingSlashAndPathWithLeadingSlash() {
+        String baseURLWithSlash = SUBMODEL_REPO_URL + "/";
+
+        assertEquals(baseURLWithSlash + SUBMODEL_REPOSITORY_PATH_WITHOUT_SLASH, SubmodelDescriptorFactory.createSubmodelRepositoryUrl(baseURLWithSlash));
```
Comment (mateusmolina-iese) [inline basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java:26]: The assertion is wrongly defined

File: basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import static org.junit.Assert.assertEquals;
+
+import org.junit.Test;
+
+public class TestSubmodelDescriptorFactory {
+
+	private static final String SUBMODEL_REPO_URL = "http://localhost:8081";
+	
+
+	private static final String SUBMODEL_REPOSITORY_PATH_WITH_SLASH = "/submodels";
```
Comment (mateusmolina-iese) [inline basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java:12]: Same issue as in the TestAasDescriptorFactory

File: basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+        String baseURLWithSlash = SUBMODEL_REPO_URL + "/";
+
+        assertEquals(baseURLWithSlash + SUBMODEL_REPOSITORY_PATH_WITHOUT_SLASH, SubmodelDescriptorFactory.createSubmodelRepositoryUrl(baseURLWithSlash));
+    }
+
+    @Test
+    public void testUrlWithoutTrailingSlashAndPathWithoutLeadingSlash() {
+        String baseURLWithoutSlash = SUBMODEL_REPO_URL;
+
+        assertEquals(baseURLWithoutSlash + SUBMODEL_REPOSITORY_PATH_WITH_SLASH , SubmodelDescriptorFactory.createSubmodelRepositoryUrl(baseURLWithoutSlash));
```
Comment (mateusmolina-iese) [inline basyx.submodelrepository/basyx.submodelrepository-feature-registry-integration/src/test/java/org/eclipse/digitaltwin/basyx/submodelrepository/feature/registry/integration/TestSubmodelDescriptorFactory.java:33]: The assertion is wrongly defined

File: basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/Helper.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+import java.net.URISyntaxException;
+
+import org.springframework.web.util.UriComponentsBuilder;
+
+
+public class Helper {
+
+	public static String createRepositoryUrl(String baseUrl, String additionalPath) {
+        try {
+            // Create a URI from the base URL
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/Helper.java:14]: Please the remove the comments

File: basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+public class TestHelper {
+
+private static final String BASE_URL = "http://localhost:8081";
+	
+	private static final String PATH = "shells";
+
+	@Test
+    public void testUrlWithTrailingSlashAndPathNameWithNoLeadingSlash() {
+		assertEquals(BASE_URL + "/" + PATH, Helper.createRepositoryUrl(BASE_URL + "/", PATH));
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java:15]: tab

File: basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+    }
+	
+    @Test
+    public void testUrlWithoutTrailingSlashAndPathNameWithNoLeadingSlash() {
+        assertEquals(BASE_URL + "/" + PATH , Helper.createRepositoryUrl(BASE_URL, PATH));
+    }
+    
+    @Test
+    public void testUrlWithTrailingSlashAndPathNameWithLeadingSlash() {
+		assertEquals(BASE_URL + "/" + PATH, Helper.createRepositoryUrl(BASE_URL + "/", "/" + PATH));
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java:25]: tab

File: basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	private static final String PATH = "shells";
+
+	@Test
+    public void testUrlWithTrailingSlashAndPathNameWithNoLeadingSlash() {
+		assertEquals(BASE_URL + "/" + PATH, Helper.createRepositoryUrl(BASE_URL + "/", PATH));
+    }
+	
+    @Test
+    public void testUrlWithoutTrailingSlashAndPathNameWithNoLeadingSlash() {
+        assertEquals(BASE_URL + "/" + PATH , Helper.createRepositoryUrl(BASE_URL, PATH));
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java:20]: make BASE_URL + "/" + PATH a private static final String EXPECTED_URL

File: basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+    @Test
+    public void testUrlWithTrailingSlashAndPathNameWithLeadingSlash() {
+		assertEquals(BASE_URL + "/" + PATH, Helper.createRepositoryUrl(BASE_URL + "/", "/" + PATH));
+    }
+	
+    @Test
+    public void testUrlWithoutTrailingSlashAndPathNameWithLeadingSlash() {
+        assertEquals(BASE_URL + "/" + PATH , Helper.createRepositoryUrl(BASE_URL, "/" + PATH));
+    }
+
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestHelper.java:32]: Add tests for urls with a contextPath already set

File: basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/RepositoryUrlHelper.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package org.eclipse.digitaltwin.basyx.core;
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/main/java/org/eclipse/digitaltwin/basyx/core/RepositoryUrlHelper.java:1]: Add license header

File: basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestRepositoryUrlHelper.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
+package org.eclipse.digitaltwin.basyx.core;
```
Comment (mateusmolina-iese) [inline basyx.common/basyx.core/src/test/java/org/eclipse/digitaltwin/basyx/core/TestRepositoryUrlHelper.java:1]: Add license header

File: basyx.aasrepository/basyx.aasrepository-feature-authorization/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/authorization/TestAuthorizedAasRepository.java [hunk 14]
```diff
@@ -529,10 +530,10 @@ public void getThumbnailWithCorrectRoleAndPermission() throws IOException {
 		
 		String accessToken = getAccessToken(DummyCredentialStore.BASYX_READER_CREDENTIAL);
 		
-		CloseableHttpResponse retrievalResponse = getElementWithAuthorization(BaSyxHttpTestUtils.getThumbnailAccessURL(createAasRepositoryUrl(aasRepositoryBaseUrl), SPECIFIC_SHELL_ID), accessToken);
+		CloseableHttpResponse retrievalResponse = getElementWithAuthorization(BaSyxHttpTestUtils.getThumbnailAccessURL(RepositoryUrlHelper.createRepositoryUrl(aasRepositoryBaseUrl, null), SPECIFIC_SHELL_ID), accessToken);
```
Comment (mateusmolina-iese) [inline basyx.aasrepository/basyx.aasrepository-feature-authorization/src/test/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/authorization/TestAuthorizedAasRepository.java:533]: Are the calls to createRepositoryUrl(...) necessary? The second par is always null. Either remove them OR remove the shells context in aasRepositoryBaseUrl and use AAS_REPOSITORY_PATH and createRepositoryUrl(...) to generate the url.

**Review by mateusmolina-iese — COMMENTED** [review]
The main issue here is that the ticket goal is not being implemented or tested. I would have expected tests for externalUrl in the aasrepository-feature-registry-integration and submodelrepository-feature-registry-integration features. Please find below the entrypoint to where the externalUrl is mapped (similarly for the submodel feature):
- https://github.com/eclipse-basyx/basyx-java-server-sdk/blob/4fc702e87e534c9723c51ac70f4484dd7c895a7d/basyx.aasrepository/basyx.aasrepository-feature-registry-integration/src/main/java/org/eclipse/digitaltwin/basyx/aasrepository/feature/registry/integration/RegistryIntegrationAasRepositoryFeature.java#L53

What happens when the / suffix is added or not added to the externalUrl there?

**Review by mateusmolina-iese — APPROVED** [review]

**Pull request merged**